src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java

```diff
@@ -501,4 +519,303 @@ public int hashCode() {
+			Geometry g = unionedPolygons.next();
+			if (!g.isEmpty()) {
+				list.add(new OGCMultiPolygon((Polygon)g, esriSR));
+			}
+
+		}
+
+		return new OGCConcreteGeometryCollection(list, esriSR);
+	}
+	
+	/**
+	 * Fixes topological overlaps in the GeometryCollecion.
+	 * This is equivalent to union of the geometry collection elements.
+	 * 
+	 * @return A geometry collection that is flattened and has no overlapping elements.
+	 */
+	public OGCConcreteGeometryCollection flattenAndRemoveOverlaps() {
+		ArrayList<Geometry> geoms = new ArrayList<Geometry>();
+		
+		//flatten and crack/cluster
+		GeometryCursor cursor = OGCStructureInternal.prepare_for_ops_(flatten().getEsriGeometryCursor(), esriSR);
+		for (Geometry g = cursor.next(); g != null; g = cursor.next()) {
+			geoms.add(g);
+		}
+		
+		//make sure geometries don't overlap
+		return removeOverlapsHelper_(geoms);
+	}
+	
+	private OGCConcreteGeometryCollection removeOverlapsHelper_(List<Geometry> geoms) {
+		ArrayList<Geometry> result = new ArrayList<Geometry>();
+		for (int i = 0; i < geoms.size(); ++i) {
+			Geometry current = geoms.get(i);
+			if (current.isEmpty())
+				continue;
+			
+			for (int j = i + 1; j < geoms.size(); ++j) {
+				Geometry subG = geoms.get(j);
+				current = OperatorDifference.local().execute(current, subG, esriSR, null);
+				if (current.isEmpty())
+					break;
+			}
+			
+			if (current.isEmpty())
+				continue;
+			
+			result.add(current);
+		}
+		
+		return new OGCConcreteGeometryCollection(new SimpleGeometryCursor(result), esriSR);
+	}
+	
+	private static class FlatteningCollectionCursor extends GeometryCursor {
+		private List<OGCConcreteGeometryCollection> m_collections;
+		private GeometryCursor m_current;
+		private int m_index;
+		FlatteningCollectionCursor(List<OGCConcreteGeometryCollection> collections) {
+			m_collections = collections;
+			m_index = -1;
+			m_current = null;
+		}
+		
+		@Override
+		public Geometry next() {
+			while (m_collections != null) {
+				if (m_current != null) {
+					Geometry g = m_current.next();
+					if (g == null) {
+						m_current = null;
+						continue;
+					}
+					
+					return g;
+				}
+				else {
+					m_index++;
+					if (m_index < m_collections.size()) {
+						m_current = m_collections.get(m_index).flatten().getEsriGeometryCursor();
+						continue;
+					}
+					else {
+						m_collections = null;
+						m_index = -1;
+					}
+				}
+			}
+			
+			return null;
+		}
+
+		@Override
+		public int getGeometryID() {
+			return m_index;
+		}
+		
+	};
+	
+	//Collectively processes group of geometry collections (intersects all segments and clusters points).
+	//Flattens collections, removes overlaps.
+	//Once done, the result collections would work well for topological and relational operations.
+	private List<OGCConcreteGeometryCollection> prepare_for_ops_(List<OGCConcreteGeometryCollection> geoms) {
```

Comment:
@stolstov This method doesn't seem to be used. The only callers I see are calling `OGCStructureInternal.prepare_for_ops_` directly.

Reply:
I've commited a change to use it. See changes to difference and intersection